cpp/src/io/parquet/reader_impl_helpers.cpp

```diff
@@ -537,6 +558,214 @@ aggregate_reader_metadata::aggregate_reader_metadata(
+                   "arrow:schema not processed.");
+    return std::nullopt;
+  }
+
+  // This lambda function converts from flatbuf::duration type to cudf::type_id
+  auto duration_from_flatbuffer = [](flatbuf::Duration const* duration) {
+    // TODO: we only need this for arrow::DurationType for now. Else, we can take in a
+    // void ptr and typecast it to the corresponding type based on the type_id.
+    auto fb_unit = duration->unit();
+    switch (fb_unit) {
+      case flatbuf::TimeUnit::TimeUnit_SECOND:
+        return cudf::data_type{cudf::type_id::DURATION_SECONDS};
+      case flatbuf::TimeUnit::TimeUnit_MILLISECOND:
+        return cudf::data_type{cudf::type_id::DURATION_MILLISECONDS};
+      case flatbuf::TimeUnit::TimeUnit_MICROSECOND:
+        return cudf::data_type{cudf::type_id::DURATION_MICROSECONDS};
+      case flatbuf::TimeUnit::TimeUnit_NANOSECOND:
+        return cudf::data_type{cudf::type_id::DURATION_NANOSECONDS};
+      default: break;
+    }
+    // 0 is simply a dummy value for the scalar
+    return cudf::data_type{};
+  };
+
+  // Question: Should we not walk the schema here and return flatbuf:Fields *
+  // and walk fields in `select_columns` function where we actually
+  // use the to_type_id function?
+
+  // Lambda function to walk a field and its children in DFS manner and
+  // return boolean walk success status
+  std::function<bool(const flatbuf::Field*)> walk_field =
+    [&schema, &duration_from_flatbuffer, &walk_field](const flatbuf::Field* field) -> bool {
+    // DFS: recursively walk over its children first
+    const auto& children = field->children();
+    if (children != nullptr) {
+      auto iter = std::find_if_not(children->cbegin(),
+                                   children->cend(),
+                                   [&walk_field](const auto& child) { return walk_field(child); });
+      if (iter != children->end()) { return false; }
+    }
+
+    // Walk the field itself
+    // TODO: is the 'name' a good key? What if cols at different levels have the same name.
+    // Solution: build strings like: "parent.child.grandchild.greatgrandchild" to ensure
+    // uniqueness?
+    auto name = (field->name() != nullptr) ? field->name()->str() : "";
+
+    if (field->type_type() == flatbuf::Type::Type_Duration) {
+      auto type_data = field->type_as_Duration();
+      if (type_data != nullptr) {
+        // add an entry to the the unordered_map
+        schema[name] = duration_from_flatbuffer(type_data);
+      } else {
+        CUDF_LOG_ERROR("Parquet reader encountered an invalid type_data pointer.",
+                       "arrow:schema not processed.");
+        return false;
+      }
+    }
+    return true;
+  };
+
+  // Walk all fields and extract all DurationType columns
+  auto iter = std::find_if_not(fields->cbegin(), fields->cend(), [&walk_field](const auto& field) {
+    return walk_field(field);
+  });
+
+  if (iter != fields->end()) { return std::nullopt; }
```

Comment:
maybe instead use `std::all_of()` here? I believe it will short-circuit.
```suggestion
  if (not std::all_of(fields->cbegin(), fields->cend(), [&walk_field](const auto& field) {
    return walk_field(field);
  })) {
    return std::nullopt;
  }
```

Reply:
I think both are semantically the [same](https://en.cppreference.com/w/cpp/algorithm/all_any_none_of#Version_1) but `std::all_of` version def looks cleaner